### ISSUE TITLE
ENH: Better support for complex numbers in scipy.signal

### ISSUE THREAD

**Issue by roryyorke** [scipy/scipy#19771]
All the below is based off commit 35f3d3f01, which is after 1.12 release branch was made; 1.12 will likely serve as a reasonable baseline for the status described below.

# Summary

This is a enhancement meta-proposal that I hope will argue the case that scipy.signal can achieve good support of complex-valued signals and complex-parameter systems with incremental changes.

The scope of what I propose is limited, and the following two goals are excluded:

  - filter design of complex-coefficient systems
  - conversion of complex-coefficient systems to second-order sections

This is a list of `scipy.signal` callables I think need changing, ordered approximately and subjectively from defect to enhancement.

| callable        | comment                                                        |
| --------------- | -------------------------------------------------------------- |
| zpk2tf          | wrong result if z, p real and k complex; no tf2zpk round-trip  |
| bilinear        | gh-13823; silent failure?                                      |
| vectorstrength  | gives weird results with complex inputs                        |
| filtfilt        | gh-17877                                                       |
| sosfiltfilt     | check gh-17877 behaviour                                       |
| coherence       | no complex unit test; needs a return\_onesided argument        |
| symiirorder1    | no unit tests. `#ifdef __GNUC__` only. See gh-18926            |
| symiirorder2    | as symiirorder1                                                |
| sepfir2d        | `#ifdef __GNUC__` only                                         |
| savgol\_filter  | complex values removed in output                               |
| spline\_filter  | forces complex to single precision; see e.g. cupy gh-7721      |
| qspline1d\_eval | perhaps a bug, given cspline1d\_eval works?                    |
| dlsim           | output set to float64 (np.zeros)                               |
| lsim            | OK if params complex, but output type set to sys.A type        |
| findfreqs       | return negative freqs for complex-coef sys                     |
| freqresp        | OK, but should auto-choose negative freqs for complex-coef sys |
| bode            | as freqresp                                                    |
| dbode           | as freqresp                                                    |
| dfreqresp       | as freqresp                                                    |
| freqz           | auto "whole=True" for complex-coef                             |
| freqz\_zpk      | as freqz                                                       |
| sosfreqz        | as freqz                                                       |
| chirp           | gh-17318; perhaps a new function complex\_chirp?               |
| sweep\_poly     | as chirp                                                       |
| lombscargle     | assume possible; don't know how difficult                      |

There are four functions I haven't checked yet that may or may not need changes:

  - cspline2d
  - qspline2d
  - check\_COLA
  - check\_NOLA

# Goals and constraints

In scipy.signal, support

1.  Complex-valued signals and
2.  Complex-parameter LTI systems (e.g., complex coef. TFs and complex SS matrices)

scipy.signal should still support real-valued signals and real-parameter systems.

## Motivation

Complex-valued signals and systems are widely used in signal processing (see, e.g., <https://github.com/scipy/scipy/pull/18883#issuecomment-1636777541>).

Such systems are also used in specialized areas in control, e.g., 3-phase inverter control. For other control applications see <https://portal.research.lu.se/portal/files/30365130/2017ACC_0939_FI.pdf>

## Backward compatibility

Many, likely the great majority, of applications expect real-valued signals and real-parameter systems. Where necessary, support for complex numbers should be explicitly opt-in.

## Non-goals

To be clear: these non-goals are **excluded** from the proposed changes.

### Extending filter design for complex-coefficients

There are papers on this, but I haven't investigated in detail at all. We could add these later as new filter design functions.

For "simple" asymmetric design one can use normal (real-coefficient, zero-frequency-centred) low- or high-pass filters and frequency-shift the poles and zeros. See e.g. <https://www.mathworks.com/help/dsp/ug/complex-bandpass-filter-design.html>

### Second-order section conversion

Converting other forms (zpk, tf, …) to second-order sections could probably be extended to support complex numbers, though it's not immediately clear how.

Arguably one only needs first-order sections if complex coefficients are allowed.

### Conversion of complex-parameter to equivalent real-parameter systems

SISO complex-parameter systems can be converted equivalent 2x2 MIMO real-parameter systems, but scipy.signal does not support MIMO systems.

# scipy.signal complex-number support status

Public callables taken from docstring of scipy.signal.\_\_init\_\_ of commit 35f3d3f01.

Support status based on my reading of code and tests, and command-line experimentation. Please correct\!

Median filters and peak finding need signal values to be ordered (that is, able-to-be-ordered; "orderable?"), and as I understand these cannot work with complex-valued signals.

## Detail

key for "status"

  - y = supported (possibly needs more unit tests)

  - c = not supported or partially supported; change to support (needs change)

  - i = impossible or unnecessary to support complex numbers (no change)

  - n = could support, but leave (no change)

  - d = deprecated (no change)

  - (blank) = unknown
    
    | callable             | status | remark                                                                     |
    | -------------------- | ------ | -------------------------------------------------------------------------- |
    | convolve             | y      |                                                                            |
    | correlate            | y      |                                                                            |
    | fftconvolve          | y      |                                                                            |
    | oaconvolve           | y      |                                                                            |
    | convolve2d           | y      |                                                                            |
    | correlate2d          | y      |                                                                            |
    | sepfir2d             | c      | `#ifdef __GNUC__` only                                                     |
    | choose\_conv\_method | i      |                                                                            |
    | correlation\_lags    | i      |                                                                            |
    | bspline              | d      |                                                                            |
    | cubic                | d      |                                                                            |
    | quadratic            | d      |                                                                            |
    | gauss\_spline        | i      |                                                                            |
    | cspline1d            | y      |                                                                            |
    | qspline1d            | y      |                                                                            |
    | cspline2d            |        | haven't figured out how to evaluate; probably should support complex?      |
    | qspline2d            |        | as cspline2d                                                               |
    | cspline1d\_eval      | y      |                                                                            |
    | qspline1d\_eval      | c      | perhaps a bug, given cspline1d\_eval works?                                |
    | spline\_filter       | c      | forces complex to single precision; see e.g. cupy gh-7721                  |
    | order\_filter        | i      | requires ordered signal values                                             |
    | medfilt              | i      | requires ordered signal values                                             |
    | medfilt2d            | i      | requires ordered signal values                                             |
    | wiener               | y      | no complex test                                                            |
    | symiirorder1         | c      | no unit tests. `#ifdef __GNUC__` only. See gh-18926                        |
    | symiirorder2         | c      | as symiirorder1                                                            |
    | lfilter              | y      |                                                                            |
    | lfiltic              | y      | only error tests                                                           |
    | lfilter\_zi          | y      | only real tests                                                            |
    | filtfilt             | c      | gh-17877                                                                   |
    | savgol\_filter       | c      | complex values removed in output                                           |
    | deconvolve           | y      | only real tests, but quick check gives good results                        |
    | sosfilt              | y      | complex type tested, but only with imag always 0                           |
    | sosfilt\_zi          | y      | as sosfilt                                                                 |
    | sosfiltfilt          | c      | check gh-17877 behaviour                                                   |
    | hilbert              | i      | inherent to the method?                                                    |
    | hilbert2             | i      |                                                                            |
    | decimate             | y      | gh-17877?                                                                  |
    | detrend              | y      | only real tests                                                            |
    | resample             | y      | tests complex type, imag=0?                                                |
    | resample\_poly       | y      |                                                                            |
    | upfirdn              | y      |                                                                            |
    | bilinear             | c      | gh-13823; silent failure?                                                  |
    | bilinear\_zpk        | y      | only tested with complex conj. pairs                                       |
    | findfreqs            | c      | in principle could handle complex LTI                                      |
    | firls                | n      | filter design                                                              |
    | firwin               | n      | filter design                                                              |
    | firwin2              | n      | filter design                                                              |
    | freqs                | y      | uses findfreqs                                                             |
    | freqs\_zpk           | y      | uses findfreqs                                                             |
    | freqz                | c      | auto "whole=True" for complex-coef                                         |
    | freqz\_zpk           | c      | as freqz                                                                   |
    | sosfreqz             | c      | as freqz                                                                   |
    | gammatone            | n      | filter design                                                              |
    | group\_delay         | n      | needs gh-19627                                                             |
    | iirdesign            | n      | filter design                                                              |
    | iirfilter            | n      | filter design                                                              |
    | kaiser\_atten        | n      |                                                                            |
    | kaiser\_beta         | n      |                                                                            |
    | kaiserord            | n      |                                                                            |
    | minimum\_phase       | n      |                                                                            |
    | savgol\_coeffs       | n      |                                                                            |
    | remez                | n      | filter design                                                              |
    | unique\_roots        | y      |                                                                            |
    | residue              | y      |                                                                            |
    | residuez             | y      |                                                                            |
    | invres               | y      |                                                                            |
    | invresz              | y      |                                                                            |
    | BadCoefficients      | i      |                                                                            |
    | abcd\_normalize      | y      |                                                                            |
    | band\_stop\_obj      | n      | filter design                                                              |
    | besselap             | n      | filter design                                                              |
    | buttap               | n      | filter design                                                              |
    | cheb1ap              | n      | filter design                                                              |
    | cheb2ap              | n      | filter design                                                              |
    | cmplx\_sort          | d      |                                                                            |
    | ellipap              | d      | filter design                                                              |
    | lp2bp                | d      | filter design                                                              |
    | lp2bp\_zpk           | d      | filter design                                                              |
    | lp2bs                | d      | filter design                                                              |
    | lp2bs\_zpk           | d      | filter design                                                              |
    | lp2hp                | d      | filter design                                                              |
    | lp2hp\_zpk           | d      | filter design                                                              |
    | lp2lp                | d      | filter design                                                              |
    | lp2lp\_zpk           | d      | filter design                                                              |
    | normalize            | y      |                                                                            |
    | butter               | n      | filter design                                                              |
    | buttord              | n      | filter design                                                              |
    | cheby1               | n      | filter design                                                              |
    | cheb1ord             | n      | filter design                                                              |
    | cheby2               | n      | filter design                                                              |
    | cheb2ord             | n      | filter design                                                              |
    | ellip                | n      | filter design                                                              |
    | ellipord             | n      | filter design                                                              |
    | bessel               | n      | filter design                                                              |
    | iirnotch             | n      | filter design                                                              |
    | iirpeak              | n      | filter design                                                              |
    | iircomb              | n      | filter design                                                              |
    | lti                  |        |                                                                            |
    | StateSpace           | y      |                                                                            |
    | TransferFunction     | y      |                                                                            |
    | ZerosPolesGain       | y      |                                                                            |
    | lsim                 | c      | OK if params complex, but output type set to sys.A type                    |
    | lsim2                | d      |                                                                            |
    | impulse              | c      | depends on lsim                                                            |
    | impulse2             | d      |                                                                            |
    | step                 | c      | uses lsim; test\_ltisys.TestStep.test\_complex\_input minimally tests step |
    | step2                | d      |                                                                            |
    | freqresp             | c      | OK, but should auto-choose negative freqs for complex-coef sys             |
    | bode                 | c      | as freqresp                                                                |
    | dlti                 |        |                                                                            |
    | dlsim                | c      | output set to float64 (np.zeros)                                           |
    | dimpulse             | c      |                                                                            |
    | dstep                | c      |                                                                            |
    | dfreqresp            | c      | as freqresp                                                                |
    | dbode                | c      | as freqresp                                                                |
    | tf2zpk               | y      | tested w/ complex128, but tf coef imag part always (nearly\!) zero         |
    | tf2sos               | n      | \*2sos tricky?                                                             |
    | tf2ss                | y      | no complex tests                                                           |
    | zpk2tf               | c      | wrong thing if k only complex                                              |
    | zpk2sos              | n      | as tf2sos                                                                  |
    | zpk2ss               | y      | no tests                                                                   |
    | ss2tf                | y      |                                                                            |
    | ss2zpk               | y      | ss2zpk not directly tested                                                 |
    | sos2zpk              | y      |                                                                            |
    | sos2tf               | y      |                                                                            |
    | cont2discrete        | y      |                                                                            |
    | place\_poles         | n      |                                                                            |
    | chirp                | c      | gh-17318; perhaps a new function complex\_chirp?                           |
    | gausspulse           | i      |                                                                            |
    | max\_len\_seq        | i      |                                                                            |
    | sawtooth             | i      |                                                                            |
    | square               | i      |                                                                            |
    | sweep\_poly          | c      | as chirp                                                                   |
    | unit\_impulse        | i      |                                                                            |
    | get\_window          | i      |                                                                            |
    | cascade              | d      |                                                                            |
    | daub                 | d      |                                                                            |
    | morlet               | d      |                                                                            |
    | qmf                  | d      |                                                                            |
    | ricker               | d      |                                                                            |
    | morlet2              | d      |                                                                            |
    | cwt                  | d      |                                                                            |
    | argrelmin            | n      | requires ordered signal values                                             |
    | argrelmax            | n      | as argrelmin                                                               |
    | argrelextrema        | n      | as argrelmin                                                               |
    | find\_peaks          | n      | as argrelmin                                                               |
    | find\_peaks\_cwt     | n      | as argrelmin                                                               |
    | peak\_prominences    | n      | as argrelmin                                                               |
    | peak\_widths         | n      | as argrelmin                                                               |
    | periodogram          | y      |                                                                            |
    | welch                | y      |                                                                            |
    | csd                  | y      |                                                                            |
    | coherence            | c      | no complex unit test; needs a return\_onesided argument                    |
    | spectrogram          | y      | has return\_onesided. No complex tests.                                    |
    | lombscargle          | c      | assume possible; don't know how difficult                                  |
    | vectorstrength       | c      | gives weird results with complex inputs                                    |
    | ShortTimeFFT         | y      |                                                                            |
    | stft                 | d      |                                                                            |
    | istft                | d      |                                                                            |
    | check\_COLA          |        | doesn't error on complex input; no idea on validity                        |
    | check\_NOLA          |        | doesn't error on complex input; no idea on validity                        |
    | czt                  | y      | no explicit complex unit test                                              |
    | zoom\_fft            | y      | has complex unit test                                                      |
    | CZT                  | y      | czt works                                                                  |
    | ZoomFFT              | y      | zoom\_fft works                                                            |
    | czt\_points          | y      |                                                                            |
    

# Related issues and PRs

  - gh-4964 signal.dlsim cannot handle complex input
  - gh-7228 scipy.signal.lsim fails if ZeroPolesGain to StateSpace produces complex matrix coefficients
  - gh-13075 BUG: fix signal.dlsim for complex input
  - gh-13823 scipy.signal.bilinear doesn't work for complex valued arrays
  - gh-17318 ENH: Add functionality of Complex Chirp waveform to Scipy.signal
  - gh-17877 BUG: filtfilt doesn't work with complex coefficient filters
  - gh-18273 BUG: dstep doesn't work with dlti with complex coefficients.
  - gh-18883 BUG: signal.dlsim: properly setting output dtype (\#13075 & \#4964)
  - gh-19627 FIX: correctly compute group\_delay for complex-coefficient TFs

**Comment by ilayn** [comment]
Thanks @roryyorke for this great summary. It must be a substantial effort to hunt these down.

We have discussed the state of `scipy.signal` offline a few times among maintainers and indeed there is some work ahead of us. Before we start adding more features to the current state we have to make sure that `scipy.signal` is actually in good shape which is unfortunately not when it comes to LTI and dLTI parts and their system functions. The algorithms are rather the naive implementations (nothing should be news to you since you are neck deep in these issues on the python-control side) and fail almost all edge cases.

We discussed adding SLICOT directly a few years back and it was not received well. Back then, I was too naive to propose it and today I agree that we don't want anymore F77 code and the binary size goes up quite a bit. And it is not a walk in the park to compile it for all platforms due to its age. In fact we are unapologetically trying to remove our F77 dependence (#18566). I am not so far from the things I'd like to clean up and I would like to get to SLICOT eventually. Once the LTI parts stabilize then we can start adding delay systems and complex-coeff systems or other obscure control theory stuff. 

Long story short, we have to go through the LTI class first and bring it to a production code state and then it will not be a massive surgery to add features to `scipy.signal`. This is probably touch all the filter design functions too and that is the part that worries me the most.

**Comment by roryyorke** [comment]
Thanks @ilayn .

> [LTI and dLTI algorithms] fail almost all edge cases

Do you have examples, perhaps pointer to a discussion on this?

One algorithm that could be improved is zpk->ss, which currently goes via TF.

Do you envision a low-level but public "algorithm" layer, operating on matrices, tf arrays, pole-zero arrays etc, and a higher-level system-focused layer?  Or only a high-level layer?

> I would like to get to SLICOT

Would this be a separate library, something like you discussed at https://github.com/python-control/python-control/discussions/928 ?

> Long story short, we have to go through the LTI class first and bring it to a production code state

OK, I understand.

> This is probably touch all the filter design functions too and that is the part that worries me the most.

Why do the filter design functions need to change - do you want them to return LTI objects?

Of the functions I listed, I think the following spectral analysis functions could be updated independently of the dLTI / LTI rework:

  - coherence
  - lombscargle
  - vectorstrength

sepfir2d could also be updated separately, though I wonder if that could be implemented via ndimage.convolve1d (I have no idea, maybe they are quite different).

symiiorder[1,2] and the spline functions have stalled PR gh-18926 to implement these i.t.o. lower-level functions, so I think it's not worth looking at complex-coefficient support for them separately.

Also independent of the dLTI and LTI changes would be adding complex forms of `chirp` and `sweep_poly`, though this is fairly esoteric.

**Comment by adammj** [comment]
I don't believe mathematically that it makes sense for lombscargle to accept complex numbers.

**Comment by FergoTheGreat** [comment]
I encountered the fact that zpk2sos does not work with complex filters. Its algorithm assumes that all complex poles and zeros must have a conjugate pair, and therefor cannot be used to generate a complex filter in SOS form.
